editor/src/clj/editor/debugging/mobdebug.clj

```diff
@@ -109,31 +201,114 @@
 
 (defrecord LuaRef [address])
 
-;; A representation of lua table that supports circular references. Acts like a
-;; read-only map (no assoc/dissoc) that creates sub-structures on demand.
+(defn- tuple->map-entry
+  ^MapEntry [[k v]]
+  (MapEntry. k v))
 
-(deftype LuaStructure [value refs]
+(defn- sequence->map-entries [coll]
+  (sequence (comp (partition-all 2)
+                  (map tuple->map-entry))
+            coll))
+
+(defn- classify-type [x]
+  (cond (number? x) :number
+        (string? x) :string
+        (instance? Comparable x) :comparable
+        :else :other))
+
+(defn- compare-keys [a b]
+  (let [a-type (classify-type a)
+        b-type (classify-type b)]
+    (case a-type
+      :number (case b-type
+                :number (compare a b)
+                (:string :comparable :other) -1)
+      :string (case b-type
+                :number 1
+                :string (compare a b)
+                (:comparable :other) -1)
+      :comparable (case b-type
+                    (:number :string) 1
+                    :comparable (compare a b)
+                    :other -1)
+      :other (case b-type
+               (:number :string :comparable) 1
+               :other (compare (System/identityHashCode a)
+                               (System/identityHashCode b))))))
+
+(defn- read-lua-table [{:keys [content string]}]
+  (with-meta (into (sorted-map-by compare-keys)
+                   (sequence->map-entries content))
+             {:string string}))
+
+(declare ->LuaStructure)
+
+(defn- decode-serialized-data
+  [debug-session edn-string]
+  (try
+    (edn/read-string
+      {:readers {'lua/table read-lua-table
+                 'lua/ref ->LuaRef
+                 'lua/structure (fn [{:keys [value refs]}]
+                                  (->LuaStructure debug-session value refs))}
+       :default ->LuaBlackBox}
+      edn-string)
+    (catch Exception e
+      (throw (ex-info "Error decoding serialized data" {:input edn-string} e)))))
+
+;; A representation of lua table that supports circular references. Acts like a
+;; read-only map (no assoc/dissoc) that creates/loads sub-structures on demand.
+
+(defn- maybe-ref->structure [^DebugSession debug-session x refs]
+  (if (instance? LuaRef x)
```

Comment:
Maybe invert this `if` clause to move up the `x` at the end of this function for clarity?